Have Legacy Mode use a Popup Window

jisqyv opened the pull request:
Starting with Chrome 94, we can no longer use Ajax requests to
communicate via http with the Companion. We also cannot communicate
with http if we are in a secure context.

This change introduces a Popup window which is loaded from the
Companion itself and can therefore communicate with the Companion. Our
main window (from replmgr.js) will now communicate with this Popup
which will act as a go between. This not only permits Chrome 94+ to
work but also permits us to load App Inventor in a secure context.

This code is both forward and backward compatible. An older Companion
will use Legacy Mode as it always has (and be subject to the
restrictions involved). Similarly the new Companion will work with an
older version of App Inventor.

The new Companion indicates that it can use a popup by sending the
"useproxy" flag to the Rendezvous server which is then queried for in
replmgr.js

Change-Id: I43005b78c190dfb8446308f692dc39dec51befb1

jisqyv commented:
Note: There is still more work to be done here (including rebasing to ucr). In particular, I need to gracefully handle the case where the popup is blocked. I am also experimenting with how old a browser can run the Popup code, given it is using newer Javascript and DOM features (aka async/await and fetch). If necessary I will transpile the code and include the actual source somewhere else in the source hierarchy, perhaps under misc along with the webpack config and package.json to build it (I am currently experimenting with npm/webpack/babel).

halatmit commented:
Nice to see that you have a solution .  Is there some way we can add a
message that will encourage people not to use an
old companion?
== Hal

Hal Abelson
***@***.***
Prof. of Comp. Sci. and Eng.
MIT Dept. of Elec. Eng. and Comp. Sci.



On Wed, Nov 24, 2021 at 12:14 PM Jeffrey Schiller ***@***.***>
wrote:

> Starting with Chrome 94, we can no longer use Ajax requests to
> communicate via http with the Companion. We also cannot communicate
> with http if we are in a secure context.
>
> This change introduces a Popup window which is loaded from the
> Companion itself and can therefore communicate with the Companion. Our
> main window (from replmgr.js) will now communicate with this Popup
> which will act as a go between. This not only permits Chrome 94+ to
> work but also permits us to load App Inventor in a secure context.
>
> This code is both forward and backward compatible. An older Companion
> will use Legacy Mode as it always has (and be subject to the
> restrictions involved). Similarly the new Companion will work with an
> older version of App Inventor.
>
> The new Companion indicates that it can use a popup by sending the
> "useproxy" flag to the Rendezvous server which is then queried for in
> replmgr.js
>
> Change-Id: I43005b78c190dfb8446308f692dc39dec51befb1
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/mit-cml/appinventor-sources/pull/2611
> Commit Summary
>
>    - c7282b9
>    <https://github.com/mit-cml/appinventor-sources/pull/2611/commits/c7282b9fce109ce7918c7b0f93fbb5682646dfdc>
>    Have Legacy Mode use a Popup Window
>
> File Changes
>
> (6 files <https://github.com/mit-cml/appinventor-sources/pull/2611/files>)
>
>    - *M*
>    appinventor/aiplayapp/src/edu/mit/appinventor/aicompanion3/Screen1.bky
>    <https://github.com/mit-cml/appinventor-sources/pull/2611/files#diff-fbbf48f71414609d958b607c2e111832ff69f8851bb7d6b57668f95590054f0f>
>    (24)
>    - *M*
>    appinventor/aiplayapp/src/edu/mit/appinventor/aicompanion3/Screen1.scm
>    <https://github.com/mit-cml/appinventor-sources/pull/2611/files#diff-44a07f18295c6f5c039cf44e222aed7bf0092fdf9323b7589ee0875a6a57849c>
>    (2)
>    - *M*
>    appinventor/aiplayapp/src/edu/mit/appinventor/aicompanion3/Screen1.yail
>    <https://github.com/mit-cml/appinventor-sources/pull/2611/files#diff-4a4bd124bc4a002424b3526203036d36cddb89deb67f5fbab698654d4e7373da>
>    (4)
>    - *M* appinventor/appengine/src/com/google/appinventor/client/Ode.java
>    <https://github.com/mit-cml/appinventor-sources/pull/2611/files#diff-8fcbd3ff952abc30ba9669440d8f1ed7513a147f60fb1006625aa8e1ba9930fb>
>    (9)
>    - *M* appinventor/blocklyeditor/src/replmgr.js
>    <https://github.com/mit-cml/appinventor-sources/pull/2611/files#diff-9d4ff6d37d5f23e45441c122c97190387cd0aad6f7929244e42fdfe7a67f9672>
>    (221)
>    - *M*
>    appinventor/components/src/com/google/appinventor/components/runtime/util/AppInvHTTPD.java
>    <https://github.com/mit-cml/appinventor-sources/pull/2611/files#diff-5c52f72219bf3b5e4ee46814ccd94fff2132be588e73ecefc41617d6cd9e4e64>
>    (110)
>
> Patch Links:
>
>    - https://github.com/mit-cml/appinventor-sources/pull/2611.patch
>    - https://github.com/mit-cml/appinventor-sources/pull/2611.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/mit-cml/appinventor-sources/pull/2611>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAGAXN3YAKVIJ7TKHNI5M3LUNUMN7ANCNFSM5IWS43EA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

jisqyv commented:
Testing: The code as is (not transpiled) works at least back to Chrome 50 (4/2016) and Firefox 52 (3/2017)

jisqyv commented:
@halatmit Most people have their Companion auto-updated from the Play Store.

We released version 2.62 on 11/9 and there are already 98,000 unique copies as seen by the Rendezvous server while the previous version (2.61) has 73,000 unique copies as soon by the Rendezvous server since 11/9. Older versions are much fewer.

halatmit commented:
Good that the autoupdating it doing well.

== Hal

Hal Abelson
***@***.***
Prof. of Comp. Sci. and Eng.
MIT Dept. of Elec. Eng. and Comp. Sci.



On Wed, Nov 24, 2021 at 1:08 PM Jeffrey Schiller ***@***.***>
wrote:

> @halatmit <https://github.com/halatmit> Most people have their Companion
> auto-updated from the Play Store.
>
> We released version 2.62 on 11/9 and there are already 98,000 unique
> copies as seen by the Rendezvous server while the previous version (2.61)
> has 73,000 unique copies as soon by the Rendezvous server since 11/9. Older
> versions are much fewer.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mit-cml/appinventor-sources/pull/2611#issuecomment-978110141>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAGAXN5A7N3GRSZOOAZT2DLUNUSPJANCNFSM5IWS43EA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

jisqyv commented:
Latest commit message isn't quite right, but the code is good.

ellelili2025 commented on appinventor/blocklyeditor/src/replmgr.js:
```diff
@@ -292,11 +292,17 @@ Blockly.ReplMgr.pollYail = function(workspace, opt_force) {
 Blockly.ReplMgr.resetYail = function(partial) {
-    top.ReplState.phoneState.initialized = false; // so running io stops
+    var rs = top.ReplState;
+    rs.initialized = false; // so running io stops
```
Is this supposed to be rs.phoneState.initialized = false ?

jisqyv replied:
Fixed

jisqyv commented:
Note: I have converted this back to a draft. It is now based on #2664 so should not be merged until after that PR is merged and this one appropriately tested and rebased.

jisqyv commented:
I've just converted this PR to a DRAFT. I'm going to refactor it into two commits. One for master, with just the website side changes, and the other for ucr with the Companion changes.

ewpatton commented on appinventor/blocklyeditor/src/replmgr.js:
```diff
@@ -1683,37 +1813,43 @@ Blockly.ReplMgr.putAsset = function(projectid, filename, blob, success, fail, fo
-        arrayview[i] = blob[i];
+    if (rs.proxy) {
+        rs.proxy.postMessage(['asset', encoder.toString(), arraybuf], rs.proxy_origin);
+        success();              // What happens if we fail?
```
Maybe document what happens if we fail.

jisqyv replied:
Actually, I'm not sure what happens. postMessage() isn't documented to return an error, though other sources reference that it can throw an exception. However, I don't think it ever will given the way we use it (no use of shared memory).

ewpatton reviewed (changes requested):
A few minor comments for us to think about.

jisqyv commented on appinventor/blocklyeditor/src/replmgr.js:
```diff
@@ -1683,37 +1813,43 @@ Blockly.ReplMgr.putAsset = function(projectid, filename, blob, success, fail, fo
 
     var conn = goog.net.XmlHttp();
     var arraybuf = new ArrayBuffer(blob.length);
+    var arrayview = new Uint8Array(arraybuf);
+    for (var i = 0; i < blob.length; i++) {
+        arrayview[i] = blob[i];
+    }
```
arrayview is a Uint8Array "view" of the arraybuf, which is used on line 1829

jisqyv commented on appinventor/blocklyeditor/src/replmgr.js:
```diff
@@ -1541,6 +1497,180 @@ Blockly.ReplMgr.getFromRendezvous = function() {
+            startlegacy();
+        }
+    }, function() {
+        return;
```
This is not an error situation. This is the promise sending a reject. This happens when we put up the companion out of date error and the user presses OK (i.e., do not continue). In this case, within the promise the connection attempt is aborted, and then it does the reject, which causes the whole process to stop.

jisqyv commented on appinventor/components/src/com/google/appinventor/components/runtime/util/AppInvHTTPD.java:
```diff
@@ -250,6 +257,14 @@ public void run() {
+      res.addHeader("Access-Control-Allow-Origin", "*");
+      res.addHeader("Access-Control-Allow-Headers", "origin, content-type");
+      res.addHeader("Access-Control-Allow-Methods", "POST,OPTIONS,GET,HEAD,PUT");
+      res.addHeader("Allow", "POST,OPTIONS,GET,HEAD,PUT");
```
I would rather not add this header on this round. Before we add this header, I would want to test it on a broad range of browsers (actually browser versions).

ewpatton commented:
I'm approving this for now, but we should definitely keep an eye on whether the header will be needed. Ideally it would be nice if we don't have to scramble.

jisqyv commented on appinventor/blocklyeditor/src/msg/ai_blockly/messages.json:
```diff
@@ -762,6 +762,9 @@
 	"Blockly.Msg.REPL_NO_LEGACY": "Legacy Connection mode does not work when MIT App Inventor is loaded over https (secure).",
 	"Blockly.Msg.REPL_SECURE_CONNECTION": "<b>Note:</b> You are on a secure connection, legacy mode on the Companion will not work",
+        "Blockly.Msg.REPL_POPUP_TITLE": "Popup Needed",
+        "Blockly.Msg.REPL_POPUP_MESSAGE": "Please click the \"Continue\" button below",
```
I may put a "Why do I have to do this" link in the dialog that will link to a page on the website. Unfortunately, exactly how you unblock pop-ups is both browser and extension dependent, and some schools may block such unblocking!
